File: frontend/src/types/studentDetail.ts
/**
 * Types for the student card (Phase 7.6b).
 *
 * Mirrors `app/schemas/student_detail.py` exactly. There is no codegen
 * step in this project, so the two files are kept in sync by hand — if a
 * field changes there, it must change here.
 */

import type { RiskTier } from "./dashboard";

/**
 * One criterion of the unit and this student's standing against it.
 *
 * The important difference from `DashboardCriterionScore`: a criterion
 * the student has NO event for is still present, with `score: null`.
 * The dashboard omits those; the card's job is showing them.
 */
export interface StudentCriterionDetail {
  criteria_id: number;
  name: string;
  category: string | null; // "attendance" | "weekly_tut" | "assessment" | "moodle"
  threshold: number;
  max_score: number;

  /** null = no mark on record. NOT the same as a mark of zero. */
  score: number | null;
  /** Only ever set for attendance / weekly_tut. */
  trend_value: number | null;

  /**
   * The normalised weekly cells behind `score`:
   *   attendance → 7 booleans        (W1–W7)
   *   weekly_tut → 6 status strings  (W2–W7)
   *
   * `null` for assessments and Moodle, which have no weekly dimension,
   * and `null` for anything ingested before 7.6b — those cells were
   * discarded at the time and there is nothing to recover them from.
   */
  weekly_values: (boolean | string)[] | null;

  recorded_at: string | null;
}

/** The three tutorial statuses, matching TUTORIAL_STATUS_CREDIT's keys. */
export type TutorialStatus = "submitted" | "late" | "not_submitted";

/**
 * One engine's independent verdict.
 *
 * ⚠️ `score` MEANS SOMETHING DIFFERENT PER ENGINE — never plot the two
 * on a shared scale. `score_kind` says which:
 *   - "badness"    (rule engine) — combined weighted badness. Higher = worse.
 *   - "confidence" (ML model)    — predicted class probability. Higher =
 *     more certain about the tier it chose, which is NOT more at risk.
 */
export interface StudentEngineDetail {
  tier: RiskTier;
  score: number;
  score_kind: "badness" | "confidence";
  is_incomplete: boolean;
  missing_criteria: string | null;
  explanation: string | null;
  computed_at: string | null;
}

/**
 * One recorded lecturer decision on an engine disagreement.
 *
 * `rule_tier` and `ml_tier` are the pair this decision was made ABOUT,
 * which is what makes carry-forward safe: a later run applies the
 * decision only if both engines still say exactly this. When they don't,
 * the card uses these to say what moved rather than silently re-asking.
 */
export interface StudentReviewDetail {
  id: number;
  decision: RiskTier;
  comment: string | null;
  rule_tier: RiskTier;
  ml_tier: RiskTier;
  reviewed_by: number;
  reviewer_name: string | null;
  created_at: string | null;
}

/** The requesting lecturer's own notes. Never another lecturer's. */
export interface StudentNoteDetail {
  body: string;
  updated_at: string | null;
}

/** The whole payload from GET /lecturer/students/{id}?unit_id=. */
export interface StudentDetailResponse {
  student_id: number;
  student_number: string;
  name: string;
  email: string | null;
  program: string | null;

  unit_id: number;
  unit_code: string;
  unit_name: string;
  enrolled_at: string | null;

  checkpoint_week: number;

  analysed: boolean;
  final_tier: RiskTier | null;
  requires_review: boolean;
  reason: string | null;
  computed_at: string | null;

  rule: StudentEngineDetail | null;
  ml: StudentEngineDetail | null;

  /** EVERY enabled criterion on the unit, including ones with no data. */
  criteria: StudentCriterionDetail[];

  note: StudentNoteDetail | null;

  /** Needed to submit a decision. null when never analysed. */
  verdict_id: number | null;

  /**
   * The review standing behind this verdict's tier — whether submitted
   * against this verdict or CARRIED FORWARD from an earlier run.
   * Non-null means a human decided this tier, not the engines.
   */
  applied_review_id: number | null;

  /** Every decision ever recorded, newest first. Append-only. */
  review_history: StudentReviewDetail[];
}

/** A lecturer's decision on an engine disagreement. */
export interface StudentReviewSubmit {
  decision: RiskTier;
  /** Optional — never required, or lecturers type "ok" fifteen times. */
  comment?: string;
}

/** Identifies which card to open. Both halves are required — risk is per unit. */
export interface StudentCardTarget {
  studentId: number;
  unitId: number;
}

File: backend/app/schemas/student_detail.py
"""
Pydantic schemas for the per-student detail card (Phase 7.6b).

WHY THIS EXISTS RATHER THAN REUSING THE DASHBOARD PAYLOAD
---------------------------------------------------------
GET /lecturer/dashboard already carries most of a student's picture, and
the students TABLE is built entirely from it. The card needs three
things that payload structurally cannot give:

  1. `RiskScore.explanation` - the rule engine's per-criterion breakdown
     and the ML model's SHAP-derived reasoning. Never sent to the
     dashboard, and not reconstructable from what is.
  2. Criteria the student has NO data for. The dashboard deliberately
     omits those (a missing mark and a zero mean different things), but
     the card's whole job is showing what is missing, so it sends every
     criterion the unit defines with a null score where there is no
     event.
  3. `weekly_values` - the per-week cells behind the attendance and
     tutorial percentages. Far too heavy to ship for a whole cohort,
     exactly right for one student.

Sending these on the dashboard payload instead would multiply its size
for every student a lecturer never opens.
"""

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel


class StudentCriterionDetail(BaseModel):
    """
    One criterion of the unit, and this student's standing against it.

    UNLIKE the dashboard's equivalent, a criterion with no recorded event
    is still returned, with `score = None`. That distinction is the point
    of the card: "no mark recorded" and "scored zero" look identical in
    an aggregate and mean completely different things to a lecturer.
    """

    criteria_id: int
    name: str
    category: Optional[str] = None
    threshold: float
    max_score: float

    # None = no AssessmentEvent exists for this student and criterion.
    score: Optional[float] = None
    # Only ever set for attendance / weekly_tut.
    trend_value: Optional[float] = None

    # The normalised weekly cells behind `score`:
    #   attendance  -> 7 booleans      (W1-W7)
    #   weekly_tut  -> 6 status strings (W2-W7)
    # None for assessments and Moodle, which have no weekly dimension,
    # and None for anything ingested before Phase 7.6b - those cells
    # were discarded at the time and cannot be recovered.
    weekly_values: Optional[list] = None

    recorded_at: Optional[datetime] = None


class StudentEngineDetail(BaseModel):
    """
    One engine's independent verdict, kept separate from the other so
    the card can show where they diverged.

    ⚠️ `score` MEANS SOMETHING DIFFERENT PER ENGINE and the two must
    never be plotted on a shared scale:
      - rule_based: the combined weighted BADNESS score. Higher = worse.
      - ml_model:   the predicted class PROBABILITY, i.e. the model's
                    confidence in the tier it chose. Higher = more sure,
                    which is not the same as more at risk.
    `score_kind` is sent so the frontend labels each one correctly
    instead of guessing from the number.
    """

    tier: str
    score: float
    score_kind: str  # "badness" | "confidence"
    is_incomplete: bool = False
    missing_criteria: Optional[str] = None
    explanation: Optional[str] = None
    computed_at: Optional[datetime] = None


class StudentReviewDetail(BaseModel):
    """
    One recorded lecturer decision on an engine disagreement.

    `rule_tier` and `ml_tier` are the pair this decision was made ABOUT.
    They are sent so the card can explain a decision that no longer
    applies - "you resolved this when the model said high risk; it now
    says low risk" - rather than silently asking again as if the
    lecturer had never decided anything.
    """

    id: int
    decision: str
    comment: Optional[str] = None
    rule_tier: str
    ml_tier: str
    reviewed_by: int
    reviewer_name: Optional[str] = None
    created_at: Optional[datetime] = None


class StudentNoteDetail(BaseModel):
    """The requesting lecturer's own notes. Never another lecturer's."""

    body: str
    updated_at: Optional[datetime] = None


class StudentDetailResponse(BaseModel):
    """Everything the student card renders, in one request."""

    student_id: int
    student_number: str
    name: str
    email: Optional[str] = None
    program: Optional[str] = None

    unit_id: int
    unit_code: str
    unit_name: str
    enrolled_at: Optional[datetime] = None

    checkpoint_week: int

    # False when the student is enrolled but the pipeline has never
    # produced a verdict for them. Every field below is None in that
    # case and the card says so plainly.
    analysed: bool = False
    final_tier: Optional[str] = None
    requires_review: bool = False
    reason: Optional[str] = None
    computed_at: Optional[datetime] = None

    rule: Optional[StudentEngineDetail] = None
    ml: Optional[StudentEngineDetail] = None

    # EVERY enabled criterion on the unit, including ones with no data.
    criteria: list[StudentCriterionDetail] = []

    note: Optional[StudentNoteDetail] = None

    # Needed to submit a decision - PATCH .../verdicts/{id}/review takes
    # a verdict id, and until 7.7 no payload the frontend received
    # carried one. None when the student has never been analysed.
    verdict_id: Optional[int] = None

    # The review STANDING BEHIND this verdict's tier, whether it was
    # submitted against this verdict or carried forward from an earlier
    # one. Non-null means a human decided this tier, not the engines,
    # and the card says so rather than letting a carried-forward
    # decision look like an automatic result.
    applied_review_id: Optional[int] = None

    # Every decision ever recorded for this student in this unit at this
    # checkpoint, newest first. Append-only, so this is the full history
    # including any the lecturer later superseded.
    review_history: list[StudentReviewDetail] = []


class StudentReviewSubmit(BaseModel):
    """A lecturer's decision on an engine disagreement."""

    decision: Literal["safe", "low_risk", "high_risk"]
    # Optional on purpose. A lecturer clearing fifteen of these will type
    # "ok" fifteen times if forced, and fifteen instances of "ok" are
    # worse than fifteen blanks - they look like justifications.
    comment: Optional[str] = None


class StudentNoteUpdate(BaseModel):
    """Request body for saving notes. Free text, deliberately."""

    body: str